### DISCUSSION TITLE
How to implement backpressure in custom input

### DISCUSSION THREAD

**Discussion by fearfate** [redpanda-data/connect#3052] [Q&A] [answered]
I'd added a `git` input, which will watch the git repository and read the files. while watching a big repository, this will going to leader to a oom crash.

**Comment by mihaitodor** [comment, marked as answer]
Hey @fearfate 👋 Hard to say without looking at the code, but, in general, the output dictates how many messages (or message batches) end up being in flight at any given time. You can also add a [`rate_limit` processor](https://docs.redpanda.com/redpanda-connect/components/processors/rate_limit/) to artificially throttle the flow. If your input buffers data preemptively, then it's up to you to make sure that buffer doesn't blow up. You could even use the API to [access](https://pkg.go.dev/github.com/redpanda-data/benthos/v4@v4.42.0/public/service#Resources.AccessRateLimit) a configured rate limit inside your input and throttle message consumption based on it, but that's probably overkill.
